diff --git a/lib/index.js b/lib/index.js
index 29575bd668634985c5894060b5a1c308611e81c4..1c6c56cdb9fbd6ab272acfde396410037531ff9a 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -1,7 +1,6 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
 exports.register = void 0;
-const JSONPath = require("jsonpath");
 const sqrl_1 = require("sqrl");
 const JSON_BRAKET_REGEX = /^\$\[([0-9]+|'(?:[^'\\]|\\.)*'|"(?:[^"\\]|\\.)*")\](.*)$/;
 function register(instance) {
@@ -42,13 +41,5 @@ function register(instance) {
         argstring: "object, path string",
         docstring: "Returns the value at the given path in the JSON object",
     });
-    instance.registerSync(function jsonPath(state, data, path) {
-        const stringified = JSON.stringify(data);
-        return JSONPath.query(JSON.parse(stringified), path);
-    }, {
-        args: [sqrl_1.AT.state, sqrl_1.AT.any, sqrl_1.AT.constant.string],
-        argstring: "object, path string",
-        docstring: "Returns the values matching the given JSONPath query",
-    });
 }
 exports.register = register;
